### PR TITLE
repair: Reduce hints and batchlog flush

### LINKED ISSUE #20259: Hints and batchlog flush takes long time to finish in repair
The hints and batchlog flush requests are issued to all nodes for each
repair request when tombstone_gc repair mode is used.

The amount of such flush requests is high when all nodes in the cluster
run repair. It is observed it takes a long time, up to 15s, for a repair
request to finish such a flush request.

### PR DESCRIPTION
**Pull request by asias**
The hints and batchlog flush requests are issued to all nodes for each repair request when tombstone_gc repair mode is used.

The amount of such flush requests is high when all nodes in the cluster run repair. It is observed it takes a long time, up to 15s, for a repair request to finish such a flush request.

To reduce overhead of the flush, each node caches the flush and only executes the real flush when some time has passed. It is safe to do so before the real flush_time is returned. Repair uses the smallest flush_time from peers as the repair time.

The nice thing about the cache on the receiver side is that all senders can hit the cache. It is better than cache on the sender side.

A slightly smaller flush_time compared to the real flush time will be used with the benefits of significantly dropped hints and batchlog flush. The tradeoff is reasonable.

Fixes #20259

Performance improvement. No backports.

### CODE REVIEW

File: repair/row_level.cc
Comment (vladzcloudius) [inline repair/row_level.cc:2295]: Please, make it configurable and LiveUpdate

**Review by vladzcloudius — CHANGES_REQUESTED** [review]
Please, make sure to allow configuring corresponding flush frequency

**Comment by vladzcloudius** [comment]
@asias could you, please, back this PR up with testing numbers: repair time before and after.
This is a performance optimization - let's see what it gives us.

File: repair/row_level.cc
Comment (asias) [reply repair/row_level.cc:2295]: It is already a config in my local version. I did not push it yet.

**Comment by asias** [comment]
Changes in 089751163767cc8b8d512873f3c85f3d0d634290:
- Added get_last_replay logic to avoid flush in case batchlog manager has replayed
- Integrated Botond's BM change.
- Added tests test_repair.py 
- Rebased to latest master

**Comment by asias** [comment]
> @asias could you, please, back this PR up with testing numbers: repair time before and after. This is a performance optimization - let's see what it gives us.

I saw 30X improvement in my test. 

    Tests: 2 nodes, with 1s batchlog delay:
    
    Before:
       Repair nr_repairs=20 cache_time_in_ms=0 total_repair_duration=40.04245328903198
    
    After:
       Repair nr_repairs=20 cache_time_in_ms=5000 total_repair_duration=1.252073049545288

File: repair/repair.cc
Comment (denesb) [inline repair/repair.cc:413]: This is unnecessary and will spam the logs during upgrade. The operator will upgrade all nodes eventually anyway.

File: db/config.cc
Comment (denesb) [inline db/config.cc:975]: The description doesn't explain what this flush request cache is.

File: repair/row_level.cc
Comment (denesb) [inline repair/row_level.cc:2353]: This is a large change. Can't we eliminate some of the churn by just flipping the if and having an early return?

**Comment by denesb** [comment]
This PR includes code I wrote, so I would like at least one other review.

File: repair/repair.cc
Comment (asias) [reply repair/repair.cc:413]: We can make it debug to make it less noisy.

File: repair/row_level.cc
Comment (asias) [reply repair/row_level.cc:2353]: If you look at the patch with --ignore-all-space, the diff will be much smaller.  Early return means we will print the finish log twice.

Comment (denesb) [reply repair/row_level.cc:2353]: Yes, I did that and it helped. Still, I usually prefer not having to do that.

**Comment by asias** [comment]
Changes in v2: 
- Updated the help message for the option
- Change one log from warning to debug

File: test/topology_custom/test_repair.py
Comment (asias) [reply test/topology_custom/test_repair.py:178]: done

File: test/topology_custom/test_repair.py
Comment (asias) [reply test/topology_custom/test_repair.py:215]: yes, done

File: repair/repair.cc
Comment (asias) [reply repair/repair.cc:427]: The participants will include the node itself at least, so it will not be empty. Also, it will be safer to set the flush time to zero by default.  If no real flush time is returned, we should not move the flush time to now.

File: repair/row_level.cc
Comment (asias) [reply repair/row_level.cc:2339]: I added the issue_flush to the log to let us know if it the flush is issued or not. It is useful to know the start and the end of the flush operation. It helped us to figure out the long flush issue in the past.

**Comment by asias** [comment]
Changes in 838b1ef4ff20a9cb3bf0f0c497a4e67f5a0f7ef9:
- Nit fix in test
- Updated the log to show if the flush is issued or not

File: db/config.cc
Comment (avikivity) [inline db/config.cc:854]: ring_delay_ms is very low in tests (0) and is mostly meaningless. I think we should redefine the timer, and set this to also be a duration.

Comment (avikivity) [reply db/config.cc:854]: Make it LiveUpdate too.

File: db/batchlog_manager.hh
Comment (avikivity) [inline db/batchlog_manager.hh:73]: Why gc_clock? It's not related to gc. Better to use lowres_clock.

**Comment by avikivity** [comment]
Looks generally good. I'd like to see a realistic measurement of the improvement (not on a laptop cluster).

File: db/batchlog_manager.hh
Comment (asias) [reply db/batchlog_manager.hh:73]: It is related to gc because we use the last replay time to decide the repair_time which is used to do gc.

File: db/config.cc
Comment (asias) [reply db/config.cc:854]: The replay is issued every 60 seconds except the first replay during boot. Only the first replay sleeps ring_delay_ms. @denesb Can you double check and confirm?

```
satic constexpr uint32_t replay_interval = 60 * 1000; // milliseconds

   auto delay = _delay;
    while (!_stop.abort_requested()) {
        try {
            co_await sleep_abortable(delay, _stop);
        } catch (sleep_aborted&) {
            co_return;
        }
        try {
            auto cleanup = post_replay_cleanup::no;
            if (++replay_counter >= _replay_cleanup_after_replays) {
                replay_counter = 0;
                cleanup = post_replay_cleanup::yes;
            }
            co_await do_batch_log_replay(cleanup);
        } catch (seastar::broken_semaphore&) {
            if (_stop.abort_requested()) {
                co_return;
            }
            on_internal_error_noexcept(blogger, fmt::format("Unexcepted exception in batchlog reply: {}", std::current_exception()));
        } catch (...) {
            blogger.error("Exception in batch replay: {}", std::current_exception());
        }
        delay = std::chrono::milliseconds(replay_interval);
    } 

```

So I think it makes sense to set  batchlog_replay_cleanup_after_replays to be a number of replay. We can make replay_interval configurable as well in a follow up if it is needed.

**Comment by asias** [comment]
Changes in d2e1b84975b903d08e355b966400b91e76cc64dc:
- Modified help message for batchlog_replay_cleanup_after_replays
- Made batchlog_replay_cleanup_after_replays liveupdate
- Rebased

**Comment by asias** [comment]
> Looks generally good. I'd like to see a realistic measurement of the improvement (not on a laptop cluster).

We need to extend existing SCT tests to have batchlog writes in the workload.  It seems we have no batchlog related tests in SCT.  @mykaul Could you help to assign?

**Comment by asias** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/batchlog_manager_test 🔹 topology_custom/test_repair ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/483/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11554/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3535/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/)
> 
> #### Failed Tests (24/36770):
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-SizeTieredCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-TimeWindowCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-SizeTieredCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-TimeWindowCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-SizeTieredCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-TimeWindowCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> 
> #### Build Details:
> * Duration: 3 hr 58 min
> * Builder: spider7.cloudius-systems.com

We need to adjust the dtest. PR is sent https://github.com/scylladb/scylla-dtest/pull/4934.

**Comment by mykaul** [comment]
> > Looks generally good. I'd like to see a realistic measurement of the improvement (not on a laptop cluster).
> 
> We need to extend existing SCT tests to have batchlog writes in the workload. It seems we have no batchlog related tests in SCT. @mykaul Could you help to assign?

@pehala ?
(I assume it's indirectly part of https://github.com/scylladb/scylla-cluster-tests/issues/8446 )

**Comment by pehala** [comment]
> @pehala ?
> (I assume it's indirectly part of [scylladb/scylla-cluster-tests#8446](https://github.com/scylladb/scylla-cluster-tests/issues/8446) )

Yup, I will add it as as a note to the issue

**Comment by asias** [comment]
Now the CI has passed and we have a plan for the sct test.  Can we get this merged? @denesb

**Comment by denesb** [comment]
> Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb

There is an unaddressed comment from @avikivity : https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161.

**Comment by asias** [comment]
> > Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb
> 
> There is an unaddressed comment from @avikivity : [#20260 (comment)](https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161).

As I mentioned above, the SCT with batchlog and repair is being worked on which will take a while. Yes, it is nice to see the big cluster results. We already have small scale tests showing significant improvement. Are we going to block the merge and wait for the SCT results? Isn't better to merge it now so it could gain more tests in our testing pipeline while waiting for the batchlog sct test.  If the batchlog sct perf test finds any issue later we can fix it as a follow up.

**Comment by mykaul** [comment]
> > > Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb
> > 
> > 
> > There is an unaddressed comment from @avikivity : [#20260 (comment)](https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161).
> 
> As I mentioned above, the SCT with batchlog and repair is being worked on which will take a while. Yes, it is nice to see the big cluster results. We already have small scale tests showing significant improvement. Are we going to block the merge and wait for the SCT results? Isn't better to merge it now so it could gain more tests in our testing pipeline while waiting for the batchlog sct test. If the batchlog sct perf test finds any issue later we can fix it as a follow up.

As long as we are careful not to merge its backports (currently we do not have the labels for it, but we could), then I think we are fine with merging this.

**Comment by denesb** [comment]
> > > Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb
> > 
> > 
> > There is an unaddressed comment from @avikivity : [#20260 (comment)](https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161).
> 
> As I mentioned above, the SCT with batchlog and repair is being worked on which will take a while. Yes, it is nice to see the big cluster results. We already have small scale tests showing significant improvement. Are we going to block the merge and wait for the SCT results? Isn't better to merge it now so it could gain more tests in our testing pipeline while waiting for the batchlog sct test. If the batchlog sct perf test finds any issue later we can fix it as a follow up.

It is possible that a change which is intended as and advertised as a performance improvement turns out to be a performance regression later. In that case, merging it has no value. That is why we ask for measurements up-front and not later.

In any case @avikivity requested the test, so I'll let him decide if he is fine with the tests being done post-merge.

**Comment by asias** [comment]
There is no update from QA on the test for a while, so I decided to hack and do the SCT test myself.

Config: 3 i3en.xlarge nodes, 10 keyspaces each has 1 table, 1800 repair requests in total.

Here is the result:

1) Repair Before: 9754s
    
f:nemesis.py l:1758 c:sdcm.nemesis p:INFO  > HJ: Finished repair on db nodes in parallel time_elapsed=9754s nr_repairs=1800 use_mgmt=False repair_ratio=2 keyspaces=['ks0', 'ks1', 'ks2', 'ks3', 'ks4', 'ks5', 'ks6', 'ks7', 'ks8', 'ks9'] 
    
2) Repair After: 752s 
    
f:nemesis.py l:1758 c:sdcm.nemesis p:INFO  > HJ: Finished repair on db nodes in parallel time_elapsed=752s nr_repairs=1800 use_mgmt=False repair_ratio=2 keyspaces=['ks0', 'ks1', 'ks2', 'ks3', 'ks4', 'ks5', 'ks6', 'ks7', 'ks8', 'ks9']

**Pull request merged**
